### PR TITLE
chore: Introduce 'yaml' library to parse YAML IaC files [CC-854]

### PR DESCRIPTION
**Pull request by ipapast**
## TL;DR
This PR introduces the “yaml” library as a robust way of parsing YAML files for the IaC flow.
It also removes the `js-yaml` library which is a transitive dependency.

### The Problem
The past few months, we have identified a few issues with the npm library we are using for parsing YAML files (`js-yaml`).
It detects both false positives and false negatives when parsing YAML files, which has not been ideal for us. We used to do workarounds to avoid the real issue, but we are now introducing support for CloudFormation (another YAML type) and it's time to address this.

#### Examples
**Example 1:**
 Library gets this string: `{{"foo": "bar"}: {"qux": "baz"}}` and it should fail as this is invalid yaml.
It does not fail, but it parses it like this: `[{"[object Object]":{"qux":"baz"}}]` which is not correct.
Runkit: https://runkit.com/embed/9tf55g330u5h

**Example 2:**
Library gets a CloudFormation yaml file with AWS intrinsic functions such as: 
 `key: !Equals [ !Ref EnvironmentName, prod ]`
This will throw an exception for unknown tag, which is fair and will fail. However this is still valid YAML.

### Options/Solutions

- We use an outdated version of `js-yaml@3.14.1`. We can upgrade to a newer version, which provides us with a functionality to extend the default schema to support tags like !Ref and !Equals etc. Example: 
```
jsYaml.CORE_SCHEMA.extend({
implicit: [],      
explicit: [
'Fn::Base64',
'Fn::Cidr',
'Fn::FindInMap',
'Fn::GetAtt',
'Fn::GetAZs',
'Fn::ImportValue',
'Fn::Join',
'Fn::Select',
'Fn::Split',
'Fn::Sub',
'Fn::Transform',
'Ref',
'Condition',
'Fn::And',
'Fn::Equals',
'Fn::If',
'Fn::Not',
'Fn::Or',
]
```
However, if we print the dependency of the package (see below), we will see that it is not only used in our code, but also in `@yarn/core`. This means that `js-yaml@3.14.1` will always be a dependency until yarn updates.

```
% npm ls --production js-yaml
snyk@1.0.0-monorepo /Users/iliannapapastefanou/Snyk-repos/snyk
├─┬ @snyk/snyk-cocoapods-plugin@2.5.2
│ └─┬ @snyk/cocoapods-lockfile-parser@3.6.2
│   └── js-yaml@3.14.1  deduped
├─┬ snyk-nodejs-lockfile-parser@1.34.2
│ ├─┬ @yarnpkg/core@2.4.0
│ │ └─┬ @yarnpkg/parsers@2.3.0
│ │   └── js-yaml@3.14.1  deduped
│ └── js-yaml@4.1.0 
└─┬ snyk-policy@1.19.0
  └── js-yaml@3.14.1
 ```
- Replacing the `js-yaml` library with a new one in the IaC code. We spiked different libraries and we have found one that it currently supports everything that we want and it's also extensible. 

### The Proposal
This PR is a proposal to replace it with a different npm library, called `yaml`. This library parses YAML contents correctly and ensures our current functionality does not break. 

**The proposed package itself is small, so we think adding this package is worth it - 91.7kB for the Minified bundle and 27kB for the Minified + Gzipped. Also, there are 0 direct dependencies. More: 
https://bundlephobia.com/result?p=yaml@1.10.2**
image

#### What are the relevant tickets?
[CC-854]

[CC-854]: https://snyksec.atlassian.net/browse/CC-854

### CODE REVIEW

File: src/cli/commands/test/iac-local-execution/file-parser.ts [hunk 1]
```diff
@@ -67,7 +67,12 @@ function parseYAMLOrJSONFileData(fileData: IacFileData): any[] {
   try {
     // the YAML library can parse both YAML and JSON content, as well as content with singe/multiple YAMLs
     // by using this library we don't have to disambiguate between these different contents ourselves
-    yamlDocuments = YAML.safeLoadAll(fileData.fileContent);
+    yamlDocuments = YAML.parseAllDocuments(fileData.fileContent).map((doc) => {
+      if (doc.errors.length !== 0) {
+        throw doc.errors[0];
```
[the remark this answers is not in the document]
Comment (ipapast) [reply src/cli/commands/test/iac-local-execution/file-parser.ts:72]: (This was to not break the existing jest tests for now)

File: src/cli/commands/test/iac-local-execution/file-parser.ts [hunk 2]
```diff
@@ -1,4 +1,4 @@
-import * as YAML from 'js-yaml';
```
Comment (gitphill) [inline src/cli/commands/test/iac-local-execution/file-parser.ts:1]: `js-yaml` was a transitive? that's a little naughty. Have we replaced it with `yaml` everywhere?

Comment (ipapast) [reply src/cli/commands/test/iac-local-execution/file-parser.ts:1]: It was, we also just recently realised this 😬 In the code, `js-yaml` is actually being imported only in this file.

Comment (gitphill) [reply src/cli/commands/test/iac-local-execution/file-parser.ts:1]: what about https://github.com/snyk/snyk/blob/master/src/lib/iac/iac-parser.ts#L2?

Comment (ipapast) [reply src/cli/commands/test/iac-local-execution/file-parser.ts:1]: 🤦 You're totally right. I just replaced it there too.

**Review by ghost — APPROVED** [review]

**Review by gitphill — APPROVED** [review]

File: src/cli/commands/test/iac-local-execution/file-loader.ts [hunk 3]
```diff
@@ -27,10 +27,14 @@ export async function loadFiles(
...
 
   const filesToScan: IacFileData[] = [];
+  const emptyFiles: string[] = [];
   for (const filePath of filePaths) {
     try {
       const fileData = await tryLoadFileData(filePath);
       if (fileData) {
+        if (fileData.fileContent === '') {
+          emptyFiles.push(fileData.filePath);
+        }
```
Comment (ipapast) [inline src/cli/commands/test/iac-local-execution/file-loader.ts:37]: I'm pushing emptyFiles in a new array, so when we have a directory containing empty files, we can compare and skip them.

File: src/cli/commands/test/iac-local-execution/file-loader.ts [hunk 4]
```diff
@@ -42,7 +46,12 @@ export async function loadFiles(
     throw new NoFilesToScanError();
   }
 
-  return filesToScan;
+  return filesToScan.filter(
+    (fileToScan) =>
+      !emptyFiles.some(
+        (emptyFilePath) => fileToScan.filePath === emptyFilePath,
+      ),
+  );
```
Comment (ipapast) [inline src/cli/commands/test/iac-local-execution/file-loader.ts:54]: @aron Can you have a look again? Now I compare the two arrays, so that we won't scan empty files at all.
I haven't added an Error for this case, but I think it can go in another PR?

Comment (aron) [reply src/cli/commands/test/iac-local-execution/file-loader.ts:54]: I don't think I understand why we even track the empty files. On line 76 of `tryLoadFileData()` we ignore files that don't have a supported extension, I think at this point, just after reading the file, we can also return null if the file is empty?

Comment (ipapast) [reply src/cli/commands/test/iac-local-execution/file-loader.ts:54]: We can do that and return null. The current flow returns an error: 
`NoFilesToScanError: Could not find any valid IaC files`
My idea was that we we skip the empty files and we let the user now which/how many files were skipped instead of an error.

**Review by aron — APPROVED** [review]
